src/Workspaces/Core/Portable/Workspace/Host/Status/ISolutionStatusService.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// Provides solution status
+    /// </summary>
+    internal interface ISolutionStatusService : IWorkspaceService
+    {
+        /// <summary>
+        /// Wait until given solution and its children are fully loaded.
+        /// </summary>
+        Task WaitForAsync(Solution solution, CancellationToken cancellationToken);
```

Comment:
WaitFor is a strange name here.  A solution is *already* a Snapshot as per how roslyn has always been documented.  So the idea that you would need to ever 'wait for' anything seems super strange.  How would one even observe any difference here?  at the very least, if waiting every changed anything, i would expect to get back the newly updated snapshot after the Wait was complete.

Reply:
if you don't like name, just propose new name. if people like it, I will take it.

Reply:
I have no idea.  The API is very strange.  I don't actually understand how it could work. Sinc ethe Solution is a snapshot... what are we waiting on?  It's not lke you could observe any difference here.  Should this instead be a Workspace API?

Reply:
good point. it should be at buffer not document.

Reply:
actually, no it is doing right thing.

using (context.OperationContext.WaitForDocumentAndAddScope(document, allowCancellation: true, EditorFeaturesResources.Applying_Extract_Method_refactoring))
            {
                return Execute(args.SubjectBuffer, args.TextView, context.OperationContext);
            }

it re-acquire document from text. so it works as expected. but, sure. the method can be change to make it more explictly.

Reply:
> actually, no it is doing right thing.

*this* specific *example* does the right thing.  My point was about the API shape in general and how this can be a pit-of-fail.  Imagine, for example, that ExtractMethod had passed the Document along.  It wouldn't work properly at all.  You would have waited for the VS solution to be fully loaded, but you'd still be accessing the document from a prior poitn in time.

That's why i'm saying we need to change teh API here.  INstead of starting with a Doc/Proj/Solution, and waiting and *then* having to reacquire.  We should allow starting *purely* from a TextBuffer or Workspace, and being able to get the Doc/Proj/Solution that iss the correct snapshot *once* the VSWorkspace has fully loaded.

Right now this API is a bit of fail because you can have multiple versions of Docs/Projs/Sols floating around, and you have to be extremely careful to make sure you're using the right one.

Reply:
well. our code already complex I am not sure what you meant by people can do things without being careful. but sure, if you have other suggestion, suggest it rather than again throwing all those words around.

Reply:
> well. our code already complex 

This is not a reason to introduce even more complexity.  Especially when tehre are potential simple alternatives we can do instead.

> I am not sure what you meant by people can do things without being careful. 

I explained it.  Imagine the code looked like this:

```c#
            using (context.OperationContext.WaitForLatestAndAddScope(document, allowCancellation: true, EditorFeaturesResources.Applying_Extract_Method_refactoring))
            {
                return Execute(document, context.OperationContext);
            }
```

This code would now be incorrect.  Because it would not be correct to use 'document' after calling `WaitForLatestAndAddScope`.  That's a pit-of-fail API.  The API makes it appear as if `document` will be ok to use after it completes.  But that's not the case.  You need to *reacquire* the document.

I've proposed several ways we could avoid this issue, and where we could ensure that the Document the person is actually using is the right one without issues.

> if you have other suggestion, suggest it rather than again throwing all those words around.

I have already created several posts suggesting alternative ways of doing things.  please read my posts.

And i'm not "throwing all these words around".  I'm taking the time to explain the problem *in depth*.  I've taken the time to give alternative approaches.   I'm spending a lot of time to help here, and it's dismissive and rude to ignore the posts that specify alternatives, then ask me for alternatives and then tell me that i'm just throwing words around as if i was not putting in a lot of effort toward getting the highest quality solution possible here.

Reply:
ya. I get that.

other approach could be WaitForLatest.. return something that one can get LatestDocument from it.

Reply:
Right.  It seems like it would make sense to be: `Task<Solution> GetFullyLoadedSolution(Workspace workspace); `.  Or, it could be `Task WaitForFullyLoadedSolution(Workspace)` and you would simply then call `.CurrentSolution` on the workspace afterwards.

Thotse both would fit far more with the Roslyn model/concepts of these entities.